core/services/ocr2/plugins/mercury/plugin_test.go

```diff
@@ -0,0 +1,282 @@
+	}
+
+	// this is kind of gross, but it's the best way to test return values of the services
+	expectedEmbeddedServiceCnt = 3
+	expectedLoopServiceCnt     = expectedEmbeddedServiceCnt + 1
+)
+
+func TestNewServices(t *testing.T) {
+	type args struct {
+		pluginConfig job.JSONConfig
+		feedID       utils.FeedID
+	}
+	tests := []struct {
+		name            string
+		args            args
+		loopMode        bool
+		wantLoopFactory any
+		wantServiceCnt  int
+		wantErr         bool
+	}{
+		{
+			name: "no plugin config error ",
+			args: args{
+				feedID: v1FeedId,
+			},
+			wantServiceCnt: 0,
+			wantErr:        true,
+		},
+
+		{
+			name: "v1 legacy",
+			args: args{
+				pluginConfig: v1jsonCfg,
+				feedID:       v1FeedId,
+			},
+			wantServiceCnt: expectedEmbeddedServiceCnt,
+			wantErr:        false,
+		},
+		{
+			name: "v2 legacy",
+			args: args{
+				pluginConfig: v2jsonCfg,
+				feedID:       v2FeedId,
+			},
+			wantServiceCnt: expectedEmbeddedServiceCnt,
+			wantErr:        false,
+		},
+		{
+			name: "v3 legacy",
+			args: args{
+				pluginConfig: v3jsonCfg,
+				feedID:       v3FeedId,
+			},
+			wantServiceCnt: expectedEmbeddedServiceCnt,
+			wantErr:        false,
+		},
+		{
+			name:     "v1 loop",
+			loopMode: true,
+			args: args{
+				pluginConfig: v1jsonCfg,
+				feedID:       v1FeedId,
+			},
+			wantServiceCnt:  expectedLoopServiceCnt,
+			wantErr:         false,
+			wantLoopFactory: &loop.MercuryV1Service{},
+		},
+		{
+			name:     "v2 loop",
+			loopMode: true,
+			args: args{
+				pluginConfig: v2jsonCfg,
+				feedID:       v2FeedId,
+			},
+			wantServiceCnt:  expectedLoopServiceCnt,
+			wantErr:         false,
+			wantLoopFactory: &loop.MercuryV2Service{},
+		},
+		{
+			name:     "v3 loop",
+			loopMode: true,
+			args: args{
+				pluginConfig: v3jsonCfg,
+				feedID:       v3FeedId,
+			},
+			wantServiceCnt:  expectedLoopServiceCnt,
+			wantErr:         false,
+			wantLoopFactory: &loop.MercuryV3Service{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.loopMode {
+				t.Setenv(string(env.MercuryPlugin.Cmd), "fake_cmd")
+				assert.NotEmpty(t, env.MercuryPlugin.Cmd.Get())
+			}
+			got, err := newServicesTestWrapper(t, tt.args.pluginConfig, tt.args.feedID)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewServices() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			assert.Len(t, got, tt.wantServiceCnt)
+			if tt.loopMode {
+				foundLoopFactory := false
+				for i := 0; i < len(got); i++ {
+					if reflect.TypeOf(got[i]) == reflect.TypeOf(tt.wantLoopFactory) {
+						foundLoopFactory = true
+						break
+					}
+				}
+				assert.True(t, foundLoopFactory)
+			}
+		})
+	}
+}
+
+// we are only varying the version via feedID (and the plugin config)
+// this wrapper supplies dummy values for the rest of the arguments
+func newServicesTestWrapper(t *testing.T, pluginConfig job.JSONConfig, feedID utils.FeedID) ([]job.ServiceCtx, error) {
```

Comment:
You might consider adding `t.Helper()` here to prevent outputing information of this function. 

https://pkg.go.dev/testing#T.Helper

Reply:
done